cores/esp8266/umm_malloc/umm_malloc.h

```diff
@@ -27,8 +35,86 @@ extern void  umm_free( void *ptr );
 
 /* ------------------------------------------------------------------------ */
 
+extern umm_heap_context_t *umm_push_heap( size_t heap_number );
+extern umm_heap_context_t *umm_pop_heap( void );
+extern umm_heap_context_t *umm_set_heap_by_id( size_t which );
+extern size_t umm_get_current_heap_id(void);
+
+
 #ifdef __cplusplus
 }
 #endif
+/*
+  This is pattered after interrupts.h
+
+  HeapSelectIram is used to temporarily select an alternate Heap.
+
+  {
+      {
+        HeapSelectIram lock;
+        // allocate memory here
+      }
+      allocations here are from the old Heap selection
+  }
+ */
+
+class HeapSelect {
+public:
+#if (UMM_NUM_HEAPS == 1)
+  HeapSelect(size_t id) { (void)id; }
+  ~HeapSelect() {}
+#else
+  HeapSelect(size_t id) : _heap_id(umm_get_current_heap_id()) {
+    umm_set_heap_by_id(id);
+  }
+
+  ~HeapSelect() {
+    umm_set_heap_by_id(_heap_id);
+  }
+
+protected:
+    size_t _heap_id;
+#endif
+};
+
+class HeapSelectIram {
+public:
+#ifdef UMM_HEAP_IRAM
+  HeapSelectIram() : _heap_id(umm_get_current_heap_id()) {
+    umm_set_heap_by_id(UMM_HEAP_IRAM);
+  }
+
+  ~HeapSelectIram() {
+    umm_set_heap_by_id(_heap_id);
+  }
+
+protected:
+    size_t _heap_id;
+
+#else
+  HeapSelectIram() {}
+  ~HeapSelectIram() {}
+#endif
+};
+
+class HeapSelectDram {
+public:
+#if (UMM_NUM_HEAPS == 1)
+  HeapSelectDram() {}
+  ~HeapSelectDram() {}
+#else
+  HeapSelectDram() : _heap_id(umm_get_current_heap_id()) {
+    umm_set_heap_by_id(UMM_HEAP_DRAM);
+  }
+
+  ~HeapSelectDram() {
+    umm_set_heap_by_id(_heap_id);
+  }
+
+protected:
+    size_t _heap_id;
+#endif
+};
```

Comment:
These are objects, they likely should be put in a separate file, e.g.: umm_malloc_scoped.h or umm_malloc_select.h or similar.

Reply:
Would `umm_heap_select.h` be acceptable?

Reply:
Moved class to umm_heap_select.h